File: clients/python/examples/confirmed_delayed.py
"""Confirmed and delayed publishing.

A confirmed publish returns the broker-assigned offset, which increases across
publishes to a partition. A delayed publish is withheld until its deadline, then
delivered like any other message.

Run a broker on 127.0.0.1:9876, then:

    uv run python examples/confirmed_delayed.py
"""

from __future__ import annotations

import asyncio
import os

from fibril import ClientOptions


async def main() -> None:
    addr = os.environ.get("FIBRIL_ADDR", "127.0.0.1:9876")
    opts = ClientOptions().with_auth("fibril", "fibril")
    async with await opts.connect(addr) as client:
        topic = "orders"
        sub = await client.subscribe(topic).prefetch(8).sub_auto_ack()
        pub = client.publisher(topic)

        first = await pub.publish_confirmed({"note": "now"})
        second = await pub.publish_confirmed({"note": "also now"})
        print(f"confirmed offsets {first} then {second} (increasing: {second > first})")

        # Withheld until the delay elapses (1 second), then delivered normally.
        await pub.publish_delayed_confirmed({"note": "later"}, 1.0)
        print("published one message delayed by 1s")

        received = 0
        async for msg in sub:
            print(f"received {msg.deserialize()!r} at offset {msg.offset}")
            received += 1
            if received == 3:
                break


if __name__ == "__main__":
    asyncio.run(main())

File: crates/client/examples/confirmed_delayed.rs
//! A confirmed publish returns an increasing offset, and a delayed publish is
//! withheld before its deadline and arrives after it.
//!
//! Run a broker on 127.0.0.1:9876 (or set FIBRIL_ADDR), then:
//!   cargo run -p fibril-client --example confirmed_delayed

use std::time::Duration;

use fibril_client::{Client, ClientOptions, FibrilError, NewMessage};

#[tokio::main]
async fn main() -> Result<(), FibrilError> {
    let addr = std::env::var("FIBRIL_ADDR").unwrap_or_else(|_| "127.0.0.1:9876".to_string());
    let client =
        Client::connect(addr.as_str(), ClientOptions::new().auth("fibril", "fibril")).await?;

    let topic = "orders";
    let mut sub = client.subscribe(topic)?.prefetch(8).sub_auto_ack().await?;
    let publisher = client.publisher(topic)?;

    let first = publisher
        .publish_confirmed(NewMessage::text("now"))
        .await?;
    let second = publisher
        .publish_confirmed(NewMessage::text("also now"))
        .await?;
    assert!(
        second > first,
        "offsets increase across confirmed publishes"
    );
    println!("confirmed offsets {first} then {second}");

    // Withheld until the delay elapses (1 second), then delivered normally.
    publisher
        .publish_delayed_confirmed(NewMessage::text("later"), Duration::from_secs(1))
        .await?;
    println!("published one message delayed by 1s");

    for _ in 0..3 {
        let msg = sub.recv().await.expect("a delivery");
        println!("received {}", msg.text()?);
    }
    Ok(())
}
